sw/device/bsp/w25q/w25q.c

```diff
@@ -0,0 +1,1247 @@
+    res = dma_launch(&trans);
+
+    // Address + Read command
+    uint32_t read_byte_cmd = ((REVERT_24b_ADDR(addr & 0x00ffffff) << 8) | FC_RD);
+    // Load command to TX FIFO
+    spi_write_word(&spi, read_byte_cmd);
+    spi_wait_for_ready(&spi);
+
+    // Set up segment parameters -> send command and address
+    const uint32_t cmd_read_1 = spi_create_command((spi_command_t){
+        .len        = 3,                 // 4 Bytes
+        .csaat      = true,              // Command not finished
+        .speed      = kSpiSpeedStandard, // Single speed
+        .direction  = kSpiDirTxOnly      // Write only
+    });
+    // Load segment parameters to COMMAND register
+    spi_set_command(&spi, cmd_read_1);
+    spi_wait_for_ready(&spi);
+
+    // Set up segment parameters -> read length bytes
+    const uint32_t cmd_read_2 = spi_create_command((spi_command_t){
+        .len        = length-1,          // len bytes
+        .csaat      = false,             // End command
+        .speed      = kSpiSpeedStandard, // Single speed
+        .direction  = kSpiDirRxOnly      // Read only
+    });
+    spi_set_command(&spi, cmd_read_2);
+    spi_wait_for_ready(&spi);
+
+    // Wait for DMA to finish transaction
+        while(!dma_is_ready());
+
+    return FLASH_OK;
+}
+
+uint8_t w25q128jw_write_standard_dma(uint32_t addr, void *data, uint32_t length) {
+    // Call the wrapper with quad = 0, dma = 1
+    return page_write_wrapper(addr, data, length, 0, 1);
+}
+
+uint8_t w25q128jw_read_quad(uint32_t addr, void *data, uint32_t length) {
+    // Sanity checks
+    if (sanity_checks(addr, data, length) == 0) return FLASH_ERROR;
+
+    // Send quad read command at standard speed
+    uint32_t cmd_read_quadIO = FC_RDQIO;
+    spi_write_word(&spi, cmd_read_quadIO);
+    const uint32_t cmd_read = spi_create_command((spi_command_t){
+        .len        = 0,                 // 1 Byte
+        .csaat      = true,              // Command not finished
+        .speed      = kSpiSpeedStandard, // Single speed
+        .direction  = kSpiDirTxOnly      // Write only
+    });
+    spi_set_command(&spi, cmd_read);
+    spi_wait_for_ready(&spi);
+
+    /*
+     * Send address at quad speed.
+     * Last byte is Fxh (here FFh) required by W25Q128JW
+    */
+    uint32_t read_byte_cmd = (REVERT_24b_ADDR(addr) | (0xFF << 24));
+    spi_write_word(&spi, read_byte_cmd);
+    const uint32_t cmd_address = spi_create_command((spi_command_t){
+        .len        = 3,                // 3 Byte
+        .csaat      = true,             // Command not finished
+        .speed      = kSpiSpeedQuad,    // Quad speed
+        .direction  = kSpiDirTxOnly     // Write only
+    });
+    spi_set_command(&spi, cmd_address);
+    spi_wait_for_ready(&spi);
+
+    // Quad read requires dummy clocks
+    const uint32_t dummy_clocks_cmd = spi_create_command((spi_command_t){
+        #ifdef TARGET_PYNQ_Z2
+        .len        = 3,               // W25Q128JW flash needs 4 dummy cycles
```

Comment:
can we put these magic numbers in defines and so in case in the future we have another flash of the same family but different specs should  be easier to port